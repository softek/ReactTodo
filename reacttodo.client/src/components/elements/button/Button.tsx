import React from 'react';
import { Button as BoostrapButton } from 'reactstrap';

export type ButtonType = 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'light' | 'dark' | 'link';

export interface ButtonProps {
  /**
   * Button label. Will be used for button content if children are not provided.
   * Always used as aria-label.
   */
  label: string;
  /**
   * Click handler
   */
  onClick?: () => void;
  /**
   * Button contents, can be any valid JSX that can be nested under a button element
   */
  children?: React.ReactNode;
  /**
   * Sets the color/type of the button.
   */
  type?: ButtonType;
}

/**
 * Button component that supports text and JSX content.
 */
export function Button({
  label, onClick, type, children,
}: ButtonProps) {
  return (
    <BoostrapButton
      color={type}
      aria-label={label}
      onClick={onClick}
    >
      {children ?? label}
    </BoostrapButton>
  );
}
