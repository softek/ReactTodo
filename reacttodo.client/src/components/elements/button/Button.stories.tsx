import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import {
  fn, within, userEvent, expect, waitFor,
} from '@storybook/test';
import { Button } from './Button';

const meta = {
  title: 'Components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
  },
  argTypes: {
    children: { control: 'text' },
  },
  args: { onClick: fn() },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

/**
 * When a label is provided but no children, the label will be used as the button content.
 */
export const LabelAsContent: Story = {
  args: {
    label: 'Button',
  },
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByRole('button', { name: /Button/i });
    await expect(button).toBeInTheDocument();
    await userEvent.click(button);
    await waitFor(() => expect(args.onClick).toHaveBeenCalled());
  },
};

/**
 * Buttons can contain any JSX content that is allowed under a button element.
 * This allows for text formatting or images to be included.
 */
export const ComplexContent: Story = {
  args: {
    label: 'Button',
    children: <strong>Bold Button</strong>,
  },
};

/**
 * There are several types of buttons available.
 */
export const Types: Story = {
  args: {
    label: 'Button',
  },
  render: () => (
    <div>
      <Button type="primary" label="primary" />
      <Button type="secondary" label="secondary" />
      <Button type="success" label="success" />
      <Button type="danger" label="danger" />
      <Button type="warning" label="warning" />
      <Button type="info" label="info" />
      <Button type="light" label="light" />
      <Button type="dark" label="dark" />
      <Button type="link" label="link" />
    </div>
  ),
};
