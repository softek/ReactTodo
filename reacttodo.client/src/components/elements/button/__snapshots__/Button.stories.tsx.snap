// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Button ComplexContent smoke-test 1`] = `
<button type="button"
        class="btn btn-secondary"
        aria-label="Button"
>
  <strong>
    Bold Button
  </strong>
</button>
`;

exports[`Components/Button LabelAsContent play-test 1`] = `
<button type="button"
        class="btn btn-secondary"
        aria-label="Button"
>
  Button
</button>
`;

exports[`Components/Button Types smoke-test 1`] = `
<div>
  <button class="btn btn-primary"
          aria-label="primary"
  >
    primary
  </button>
  <button class="btn btn-secondary"
          aria-label="secondary"
  >
    secondary
  </button>
  <button class="btn btn-success"
          aria-label="success"
  >
    success
  </button>
  <button class="btn btn-danger"
          aria-label="danger"
  >
    danger
  </button>
  <button class="btn btn-warning"
          aria-label="warning"
  >
    warning
  </button>
  <button class="btn btn-info"
          aria-label="info"
  >
    info
  </button>
  <button class="btn btn-light"
          aria-label="light"
  >
    light
  </button>
  <button class="btn btn-dark"
          aria-label="dark"
  >
    dark
  </button>
  <button class="btn btn-link"
          aria-label="link"
  >
    link
  </button>
</div>
`;
